### ISSUE TITLE
What about Risc-V support future ?

### ISSUE THREAD

**Issue by eix128** [DynamoRIO/dynamorio#3544]
Hi , as we all know , Risc-V popularity is rising more in time.
Will Any support be avalible for Risc-V ?

There is already new soc released for risc-v

https://www.sifive.com/boards

**Comment by AssadHashmi** [comment]
Hi, we don't currently have plans for RISC-V support.
However, we do support AArch32 and AArch64 which can be used as reference points for a RISC-V port:
https://github.com/DynamoRIO/dynamorio/wiki/ARM-Port
https://github.com/DynamoRIO/dynamorio/wiki/AArch64-Port

We are a small developer team but welcome and will support anyone or any group who want to add features such as RISC-V. Do you have specific use-cases in mind, i.e. what sort of problems would you hope to address with DynamoRIO if RISC-V was available?

**Comment by AssadHashmi** [comment]
As an aside, I would suggest posting your question on the users list https://groups.google.com/forum/#!forum/DynamoRIO-User

This will be seen by more people than this issues list and we'll get an idea of how many are interested in RISC-V support.

**Comment by lazyparser** [comment]
Hi,

Just FYI that the PLCT Lab is planning to port the DynamoRIO to RISC-V (RV64G)[1]. Not started yet, but we are optimistic that we are able to finish the porting project and contributing back to upstream before the end of 2021.

[1] https://github.com/plctlab/plctlab.github.io/blob/main/PLCT-Roadmap-2021.md#enable-dynamorio-running-on-rv64gc

**Comment by derekbruening** [comment]
That's exciting news!

I have a few suggestions:
+ Please consider a workflow where code changes are sent for review early and often, keeping reviews small and incremental.  This makes the review process much more manageable.  (What we would like to avoid is building up a very large parallel forked code base with no review or merge until it is too large to practically review.)  Initially, existing project members would perform reviews, but once PLCT developers are up to speed they can be added as project members and perform reviews.
+ Naturally we would like to have automated testing of RISC-V added to the Continuous Integration suite: https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#automated-test-machines.  This could be via Jenkins on a machine hosted somewhere else, or through a Github Actions-provided VM if that can be configured properly.

**Comment by lazyparser** [comment]
Hi @derekbruening 

Thanks for the suggestions. We will send patch set for early review/comments once we got a basic framework. The process of porting is also open (on github, as a fork of the repo).

> Naturally we would like to have automated testing of RISC-V added to the Continuous Integration suite: https://github.com/DynamoRIO/dynamorio/wiki/Test-Suite#automated-test-machines. This could be via Jenkins on a machine hosted somewhere else, or through a Github Actions-provided VM if that can be configured properly.

Thanks. I'll check it out. I've built some CI infrastructure before, so I am optimistic that automated testing of DynamoRIO might not be hard to me:)

**Comment by onezibo** [comment]
Hi @lazyparser 

Can I know your current progress? I am currently preparing to implement RISC-V's DBI tool base on dynamorio, I hope to communicate with you more in the future. I am currently a graduate student at iscas.

Thanks,
Onezibo

**Comment by lazyparser** [comment]
> Hi @lazyparser
> 
> Can I know your current progress? I am currently preparing to implement RISC-V's DBI tool base on dynamorio, I hope to communicate with you more in the future. I am currently a graduate student at iscas.
> 
> Thanks,
> Onezibo

Hi @onezibo 

Glad to hear you are working on this. I had not start to coding yet. The plan is to finish the porting/support before the end of year 2021, and get started after April.

Feel free to drop me an email. :-)

**Comment by derekbruening** [comment]
> The process of porting is also open (on github, as a fork of the repo).

Somewhat echoing the earlier suggestions, I would suggest *not* having a forked repo being the official port: it will diverge and we'll end up with a fractured state where RISC-V is only available with a stale and outdated rest of the system.  A forked repo results in a tendency to commit only to that repo, to build up large differences which are impractical to review and merge, and combined with ongoing evolution of the other repo eventually results in permanent divergence.  IMHO it is better to build new features directly in the community's existing repo from the beginning; it is always more work to try to merge later.

**Comment by lazyparser** [comment]
> > The process of porting is also open (on github, as a fork of the repo).
> 
> Somewhat echoing the earlier suggestions, I would suggest _not_ having a forked repo being the official port: it will diverge and we'll end up with a fractured state where RISC-V is only available with a stale and outdated rest of the system. A forked repo results in a tendency to commit only to that repo, to build up large differences which are impractical to review and merge, and combined with ongoing evolution of the other repo eventually results in permanent divergence. IMHO it is better to build new features directly in the community's existing repo from the beginning; it is always more work to try to merge later.

@derekbruening I totally agree with you. The fork under the github/plctlab account is only used for sending pull requests to upstream. My long term goal is to make the RISC-V as a tier-1 platform support in upstream.

**Comment by derekbruening** [comment]
> @derekbruening I totally agree with you. The fork under the github/plctlab account is only used for sending pull requests to upstream. My long term goal is to make the RISC-V as a tier-1 platform support in upstream.

+1, SGTM!

**Comment by bekcpear** [comment]
Hi, I started porting DynamoRIO to RISC-V and it's just starting. At present, I directly use the relevant code of AArch64 to make a general framework, and then I will solve the ISA-related problems first, and make the overall RISC-V framework clearer and easier for more developers to participate.

All work will be displayed under **the default branch** of this repository: https://github.com/bekcpear/dynamorio

I will synchronize the follow-up progress here in time, and I also very much hope that you can give me some suggestions on porting or codes.

**Comment by semihalf-kardach-stanislaw** [comment]
Hi everyone.
What a lovely coincidence as I have recently started working on DynamoRIO on RISC-V as well. @bekcpear it would be awesome if we could synchronize our work and collaborate.

So far I have a codebase that can be fully compiled and linked (both cross and native) which contains mostly non-functional stubs and some trivial bits (some possibly wrong). The biggest missing piece is the instruction encoder/decoder, which is next on my TODO list.

The code can be found in `experimental-riscv` branch of the following repo: https://github.com/Semihalf/dynamorio. This will also be my main development branch so You can expect updates.

At this point I would appreciate any feedback on the porting approach I've taken (which is also heavily based on the Aarch64) port. I think we shuold start discussing on a possible changes to architecture specific code separation as the current `#ifdef` may lead to unreadable code.

Additionally there are two commits where I've tried to separate specific questions that I have:
1. [7992b2e7 RFC: Use semantic register names to simplify code](https://github.com/Semihalf/dynamorio/commit/7992b2e7d4ac84f0d769ce87a5f8a58c0c01ef59) - Postulate to designate semantic register names to simplify `IF_<arch>_ELSE(...)` ladders which will complicate code layout once a new architecture is added.
2. [38e3e239 RFC: Add comments for upstream discussion](https://github.com/Semihalf/dynamorio/commit/38e3e239f60dfa8d8a94318cc6221dd97fb15d4b) - Several questions on implementation details such as flags register assumptions, use of semantic register names or streamlining some debug tracing.

**Comment by derekbruening** [comment]
Please see our suggestions and requests above, especially https://github.com/DynamoRIO/dynamorio/issues/3544#issuecomment-822003110 and https://github.com/DynamoRIO/dynamorio/issues/3544#issuecomment-766147223: we do *not* recommend separate forks.  

Building up code in separate repositories makes it much harder to merge later.  It happens all too often that it is so much work to merge a huge amount of code that was developed separately that it never happens.  I am afraid we will end up with multiple conflicting ports that will all drift from the main repository and never be merged...it is much less work in the long run to work in the main repository from the start and most importantly **send small incremental code reviews early and often**.

**Comment by abhinav92003** [comment]
+1 to Derek's suggestion. Additionally, forked repositories do not automatically get the bug fixes and features that we regularly add to the main repository, which accumulate over time. It is also easier to report/debug issues and get feedback on new code if it is merged with the main repository.

Echoing https://github.com/DynamoRIO/dynamorio/issues/3544#issuecomment-766147223, new developers can be provided more privileges when they are up to speed, so that they can help perform code reviews for their groups. We also have a [Triager rotation](https://dynamorio.org/page_triager.html) in place to provide better support to our community, for code reviews and [dynamorio-users](https://groups.google.com/g/dynamorio-users) list discussions.

**Comment by bekcpear** [comment]
Hi, all. I'm glad to receive your attention. I really want to merge RISC-V support into this repository, not separate development in my forked repo (the task of my fork is to make PR). I'm still familiar with the overall structure, and I will submit an appropriate PR as soon as possible.

I am also very much looking forward to collaborating with you ;) @semihalf-kardach-stanislaw. If you get some work done, hopefully it can be updated instantly (new commits on your fork or make a new PR to this repo) to avoid duplication, so do I.

**Comment by semihalf-kardach-stanislaw** [comment]
I think there I've caused some confusion by being brief. The reason I've posted my code is to start the integration process as soon as possible (hence the fork as a pull request source).

I haven't posted anything earlier because achieving full compilation would help finding places that need to be stubbed/implemented and at the same time it would provide some semblance of validation (even though not a runtime one) that what I'm doing makes sense. @derekbruening Do you consider such pull request too big and therefore I should split it?

Second reason is that I want to keep my development process open and visible, especially inbetween pull requests. This is so that I can discuss with You whether my understanding of a given module is proper or not, all without spamming RFC pull requests. In essence I want to shorten the feedback loop and allow a broader community to chime in.

So my current plan is the following:

1. Sync with @bekcpear to prepare the initial pull request content.
2. Start a discussion on [dynamorio-users](https://groups.google.com/g/dynamorio-users) about potential re-factoring of the code to better accomodate 3 concurrent architectures.
3. Keep working on adding new features incrementally.
4. Figure out how to add RISC-V to CI - Qemu or physical U740 machines, how to connect them securely, etc.

Please let me know if that makes sense to everyone.

**Comment by lazyparser** [comment]
> Figure out how to add RISC-V to CI - Qemu or physical U740 machines, how to connect them securely, etc.

I'd like to contribute here: either Cloud VM (amd64, hosting QEMU/RISC-V) or physical Unmatched are available, provided by the PLCT Lab. Collaborating with RISC-V International, the PLCT Lab, ISCAS has built a large RISC-V Lab for open source communities. 

Feel free to send a PR to https://github.com/plctlab/riscv-lab-access , appending your github-id to the developers.list (the system will grab your pubkeys from github).

**Comment by lazyparser** [comment]
> Thank you for the offer! For connecting to Github, Jenkins may be a good fit: we're currently using that to run the test suite on an AArch64 machine for pull requests.

Is there any examples or docs (URL) for this steps? the machines are ready, and hopefully I have enough time to deploy the CI this weekend.

**Comment by derekbruening** [comment]
> > Thank you for the offer! For connecting to Github, Jenkins may be a good fit: we're currently using that to run the test suite on an AArch64 machine for pull requests.
> 
> Is there any examples or docs (URL) for this steps? the machines are ready, and hopefully I have enough time to deploy the CI this weekend.

@AssadHashmi set up the AArch64 Jenkins: Assad, were there some docs that you followed?

**Comment by AssadHashmi** [comment]
> @AssadHashmi set up the AArch64 Jenkins: Assad, were there some docs that you followed?

I don't have any formal documentation but this is what I did, in summary:

1. Installed Jenkins as described here https://www.jenkins.io/doc/book/installing/linux/
2. Installed '_GitHub Pull Request Builder_' plugin, see https://plugins.jenkins.io/ghprb/
3. Created a pipeline on Jenkins server and made it run our pre-commit tests. When creating '_New item_', (i.e. job) I used '_Freestyle project_' not '_Pipeline_'.
4. Created credentials for GitHub user `drbot0`.
5. Tested connection from Jenkins to GitHub using the '_Test Credentials_' feature of the '_GitHub Pull Request Builder_' section in Jenkins system configuration. This uses the existing `drbot0` user and its Personal Access Token (PAT).
6. Configured GitHub to trigger our Jenkins tests by adding a Webhook URL pointing to our Jenkins '_GitHub Pull Request Builder_' (ghprbhook), see https://github.com/DynamoRIO/dynamorio/settings/hooks
7. Tested manually from a GitHub PR using a comment phrase which can be configured in Jenkins instance. We use `run arm tests`. Got someone else to do the same making sure that person's user name was in the whitelist of our Jenkins pre-commit pipleine.

**Comment by kobrineli** [comment]
@shiptux @dolanzhao @derekbruening @semihalf-kardach-stanislaw 

Hi! What is the current state of the RISC-V support?
I've tried to build it under the Qemu for riscv64, but drrun failed on the simpliest hello world program.
The debug version printed this: `ERROR: unable to inject: exec of |(null)| failed`

**Comment by shiptux** [comment]
> @shiptux @dolanzhao @derekbruening @semihalf-kardach-stanislaw 
> 
> Hi! What is the current state of the RISC-V support?
> I've tried to build it under the Qemu for riscv64, but drrun failed on the simpliest hello world program.
> The debug version printed this: `ERROR: unable to inject: exec of |(null)| failed`

Hi thank you for your attention.
The current status of DynamoRIO RISC-V port is stage 1 (could be compiled but couldn't run). Many instruction encoder are not-implement. This also the reason why the simplest hello world program failed.

For your question, we will need instruction encoder to get it execute.
I made the DynamoRIO RISC-V units test compile last month, and then we should start the most basic instructions support.

Thanks

**Comment by kobrineli** [comment]
Got it

Thanks!

**Comment by derekbruening** [comment]
There is a race in the RISC-V build config: #6140

**Comment by ksco** [comment]
Hi,
If anyone is interested, I have successfully run a simple hello world program using `drrun` on a RISC-V SBC.

```bash
debian@lpi4a:/media/debian/dynamorio/build$ cat hello.S
.global _start
_start: li      a0, 1
        la      a1, helloworld
        li      a2, 14
        li      a7, 64
        ecall

        li      a0, 0
        li      a7, 93
        ecall

.data
helloworld:     .ascii "Hello, World!\n"

debian@lpi4a:/media/debian/dynamorio/build$ ./hello
Hello, World!
debian@lpi4a:/media/debian/dynamorio/build$ bin64/drrun hello
<Starting application /media/debian/dynamorio/build/hello (36030)>
<Initial options = -no_dynamic_options -code_api -stack_size 56K -signal_stack_size 32K -max_elide_jmp 0 -max_elide_call 0 -early_inject -emulate_brk -no_inline_ignored_syscalls -native_exec_default_list '' -no_native_exec_managed_code -no_indcall2direct >
<Paste into GDB to debug DynamoRIO clients:
set confirm off
add-symbol-file '/media/debian/dynamorio/build/lib64/debug/libdynamorio.so' 0x0000003fee3c70f0
>
Hello, World!
<Stopping application /media/debian/dynamorio/build/hello (36030)>
debian@lpi4a:/media/debian/dynamorio/build$
```

Next, I will split this work into smaller patches and submit them upstream.

**Comment by ksco** [comment]
Update the newest progress on DynamoRIO RISCV64:

Single-threaded Linux common commands can be run using drrun!

<img width="874" alt="image" src="https://github.com/DynamoRIO/dynamorio/assets/15257961/0de0b43f-14ee-4439-adf0-d654582b1007">

Also, before merged into upstream (which takes time), you can check out the progress at: https://github.com/ruyisdk/dynamorio

**Comment by mariospaok4** [comment]
Hello everyone,

We have found out that the dcontext struct is too big for all its members to be addressable using a single ld/st command, because the immediate part overflows. For that reason functions like instr_create_restore_from_dc_via_reg, dr_insert_read_tls_field etc generate broken code. dr_insert_read_tls_field, for example, generates a ld instruction with a 0x820 offset, which overflows to -0x7E0.

In order to fix that, we intend to work on reducing the size of this struct, by storing a pointer for the vector registers in the mcontext struct (instead of an array, which is the current implementation). As a side effect, this will allow detecting the size of the vector registers at runtime, and support arbitrary sizes in the future (the current DdynamoRIO supports up to 256 bit registers, which way bellow the 64K bits limit the RISCV vector spec sets).

Since the code required for that is not trivial, we would like some feedback before starting working on this patch. Are there any comments on this approach? Are there any objections, that could prevent the patch from getting merged?

**Comment by derekbruening** [comment]
It seems there are 2 separate issues here: one is the base-plus-offset limit being hit on RISC-V.  Solving that does not require changing the mcontext layout: the mcontext could still all be one single inline struct.  It does presumably require an extra scratch register or modifying the base register on RISC-V which seems like a significant change but likely there is no way around it.  Such a change would probably be RISC-V only to avoid needing more scratch registers on other platforms?

The other issue is whether the mcontext layout should change.  If your proposal is to allocate the SIMD fields on the heap with the mcontext struct pointing to the heap allocation, this needs careful consideration as there are signal handler paths where  we need to declare and use new mcontexts and allocating heap is problematic.  If heap is required to make any new mcontext, all those paths need to be carefully examined and a solution created: signal-safe heap or something, which has provisioning overheads and downsides.

The size of the mcontext is a problem on other platforms as well, since DR uses small thread stacks.  Allocating several mcontexts across a call chain can overflow the stack and with today's large SIMD fields inside mcontext we have to be careful declaring them on the stack.  See comments in signal.c like these:
```C
    /* It's safe to allocate since we do not send signals that interrupt DR.
     * With priv_mcontext_t x2 that's a little big for stack alloc.
     */
    si.mcontext = heap_alloc(dcontext, sizeof(*si.mcontext) HEAPACCT(ACCT_OTHER));
    si.raw_mcontext = heap_alloc(dcontext, sizeof(*si.raw_mcontext) HEAPACCT(ACCT_OTHER));
```
```C
/* Helper that takes over the current thread signaled via SUSPEND_SIGNAL.  Kept
 * separate mostly to keep the priv_mcontext_t allocation out of
 * main_signal_handler_C.
```
So if we do change the mcontext layout, it is worth seriously considering a cross-platform design change and not just think about RISC-V for moving some or all of mcontext to the heap.

On that same note, AArch64 also has variable-sized SIMD, and the current mcontext has a hardcoded limit (`We currently support 512-bit maximum due to DR's stack size limitation`), so the same dynamic support for larger SIMD via indirection would benefit that arch as well: so again I would make any mcontext layout change to use indirection a cross-platform design.

**Comment by ziyao233** [comment]
> It does presumably require an extra scratch register or modifying the base register on RISC-V which seems like a significant change but likely there is no way around it. Such a change would probably be RISC-V only to avoid needing more scratch registers on other platforms?

I suggest to make the base register pointing to the middle of dcontext_t instead of the start. Making use of the ability to index negatively against the base register doubles the addressable space and we won't lose another available register.

We should also check for immediate overflow when encoding load/store pairs in a debug build, like what has been done for `auipc`.

**Comment by derekbruening** [comment]
I filed #7212 on separating out `dr_simd_t`.

For having a pointer to the middle: @abhinav92003 also proposed that internally.  The offsets reach back to the start of the mcontext because dr_simd_t is limited to 256-bit registers?  The mcontext must be pretty close to the 2047 max reach; the rest of the dcontext isn't all that large but it must just push it over when reaching from the start.

You would have to audit the uses in all the generated code: there are cases where it is passed as a parameter to a function.  E.g., fcache_return passes it to d_r_dispatch; there may be others.  These would have to be handled somehow.

**Comment by mariospaok4** [comment]
Hello everyone,

It looks like #7212 is a long term endeavour, because of various aspects that have to be considered. Also it is necessary only for implementations with large vector registers. On the other hand the offset problem seems more urgent, because it affects all RISC-V implementations.

For that problem, @ziyao233's proposal will work, but we will have to make sure that we convert the dcontext pointer correctly in the entire code base. I was thinking something like adding 0x800 when storing the pointer to the TLS and subtracting 0x800 when restoring it for non instrumentation usage. People more familiar than me on the code base can provide some feedback on this idea.

Another way around that, that will need some minor changes to the shared code, is allowing functions like `instr_create_restore_from_dcontext` to have variants that return multiple instructions by reference. Then we can inject the required addi instructions before the ld/st instruction. This will not require using extra registers, since we can use the output register for the intermediate values. Compared to @ziyao233's proposal, it has the overhead of emitting an extra instruction but it will work if we need a bigger than 4KB dcontext in the future. I also do not know how intrusive the required changes will be, compared to the above proposal.

We can work on either of the above solutions. What is the preferred way?

Finally, I agree with checking for immediate overflow during encoding. Especially since it does not seem to require a lot of work to implement.

**Comment by derekbruening** [comment]
For the mid-dcontext-pointer approach: it may simplify things to add a new TLS slot that holds the mid-dcontext-pointer and then you don't need to change uses that want the base pointer only: such as get_thread_private_dcontext() which is called in many, many places.  I.e., there are 2 TLS slots for dcontext for RISCV: the current base pointer plus a new one holding the mid-pointer.  Things like fcache_return will have an extra load on RISCV to get the base pointer at the end to pass to dispatch.
